### PR TITLE
Support dynamic check podSecurityPolicy support

### PR DESCRIPTION
**Pull request by timdengyun**
Since PodSecurityPolicy is not supported any longer starting k8s >= v1.25.0,
So, it's good that using runtime k8s version decides to apply
PodSecurityPolicy or not. If k8s >= v1.25.0, PodSecurityPolicy resources
will not be applied.

### CODE REVIEW

File: pkg/controller/configmap/configmap_controller.go [hunk 1]
```diff
@@ -756,3 +775,29 @@ func secretNotEqual(s1 *corev1.Secret, s2 *corev1.Secret) bool {
 	}
 	return compareKey(s1, s2, "tls.crt") || compareKey(s1, s2, "tls.key") || compareKey(s1, s2, "tls.ca")
 }
+
+func podSecurityPolicySupport(c *rest.Config) bool {
```
Comment (dantingl) [inline pkg/controller/configmap/configmap_controller.go:779]: Do you think we can check K8s version only once when NCP Operator starts, or you think K8s may do upgrade during operator runtime?

File: pkg/controller/configmap/configmap_controller.go [hunk 2]
```diff
@@ -492,6 +499,16 @@ func (r *ReconcileConfigMap) Reconcile(request reconcile.Request) (reconcile.Res
 			}
 		}
 
+		// Since PodSecurityPolicy is not supported any longer starting k8s >= v1.25.0,
+		// So, it's good that using runtime k8s version decides to apply PodSecurityPolicy or not
+		if obj.GroupVersionKind().Kind == "PodSecurityPolicy" {
+			if !podSecurityPolicySupport(r.config) {
+				err := errors.New("PodSecurityPolicy support check failed")
+				log.Error(err, "PodSecurityPolicy is not supported from k8s >= v1.25.0")
```
Comment (dantingl) [inline pkg/controller/configmap/configmap_controller.go:507]: How about a warning or even debug log

File: pkg/controller/configmap/configmap_controller.go [hunk 3]
```diff
@@ -756,3 +775,29 @@ func secretNotEqual(s1 *corev1.Secret, s2 *corev1.Secret) bool {
...
+		return false
+	}
+
+	serverVersion, err := clientset.Discovery().ServerVersion()
+	if err != nil {
+		log.Error(err, "failed to get running Kubernetes version")
+		return false
+	}
+
+	runningVersion, err := version.ParseGeneric(serverVersion.String())
```
Comment (dantingl) [inline pkg/controller/configmap/configmap_controller.go:794]: What's the difference between serverVersion and runningVersion?

File: pkg/controller/configmap/configmap_controller.go (hunk 1 above)
Comment (timdengyun) [reply pkg/controller/configmap/configmap_controller.go]: Thanks for advice. Putting k8s version check in SharedInfo during startup initialization. So there is only one time effort to check k8s version and record result into a flag.

File: pkg/controller/configmap/configmap_controller.go (hunk 2 above)
Comment (timdengyun) [reply pkg/controller/configmap/configmap_controller.go]: Thanks. changing it to debug level log

File: pkg/controller/configmap/configmap_controller.go (hunk 3 above)
Comment (timdengyun) [reply pkg/controller/configmap/configmap_controller.go]: Semtically, we can say serverVersion is the same thing with runningVersion, just different representation. 
serverVersion is a var with data type struct, which includes more detailed version info. 
type Info struct {
	Major        string `json:"major"`
	Minor        string `json:"minor"`
	GitVersion   string `json:"gitVersion"`
	GitCommit    string `json:"gitCommit"`
	GitTreeState string `json:"gitTreeState"`
	BuildDate    string `json:"buildDate"`
	GoVersion    string `json:"goVersion"`
	Compiler     string `json:"compiler"`
	Platform     string `json:"platform"`
}

runningVersion is parsed from serverVersion's GitVersion that is a "generic" version string, the format is like
v.1.25 or 1.25. runningVersion is an opqaue representation of a version number

type Version struct {
	components    []uint
	semver        bool
	preRelease    string
	buildMetadata string
}

The reason why we convert serverVersion.string to runningVersion is that we could type Version's provide an unified interface function to handle version info regardless of platform.

File: pkg/controller/sharedinfo/shared_info.go [hunk 4]
```diff
@@ -78,5 +82,39 @@ func New(mgr manager.Manager, operatorNamespace string) (*SharedInfo, error) {
 	if ncpInstall.Spec.AddNodeTag == false {
 		addNodeTag = false
 	}
-	return &SharedInfo{AdaptorName:adaptorName, AddNodeTag:addNodeTag}, nil
+
+	podSecurityPolicySupport := false
```
Comment (dantingl) [inline pkg/controller/sharedinfo/shared_info.go:86]: podSecurityPolicySupport := isPodSecurityPolicySupport(mgr.GetConfig())

File: pkg/controller/sharedinfo/shared_info.go [hunk 5]
```diff
@@ -78,5 +82,39 @@ func New(mgr manager.Manager, operatorNamespace string) (*SharedInfo, error) {
...
+		PodSecurityPolicySupport: podSecurityPolicySupport,
+	}, nil
+}
+
+func isPodSecurityPolicySupport(c *rest.Config) bool {
+	version125, _ := version.ParseGeneric("v1.25.0")
+
+	clientset, err := clientset.NewForConfig(c)
+	if err != nil {
+		log.Error(err, "failed to creat clientset")
```
Comment (dantingl) [inline pkg/controller/sharedinfo/shared_info.go:102]: create

File: pkg/controller/sharedinfo/shared_info.go [hunk 6]
```diff
@@ -78,5 +82,39 @@ func New(mgr manager.Manager, operatorNamespace string) (*SharedInfo, error) {
...
+		return false
+	}
+
+	serverVersion, err := clientset.Discovery().ServerVersion()
+	if err != nil {
+		log.Error(err, "failed to get running Kubernetes version")
```
Comment (dantingl) [inline pkg/controller/sharedinfo/shared_info.go:108]: running -> server?

File: pkg/controller/sharedinfo/shared_info.go (hunk 4 above)
Comment (timdengyun) [reply pkg/controller/sharedinfo/shared_info.go]: Done

File: pkg/controller/sharedinfo/shared_info.go (hunk 6 above)
Comment (timdengyun) [reply pkg/controller/sharedinfo/shared_info.go]: Done

File: pkg/controller/sharedinfo/shared_info.go (hunk 5 above)
Comment (timdengyun) [reply pkg/controller/sharedinfo/shared_info.go]: Done

**Review by dantingl — APPROVED** [review]

**Pull request merged**